### DISCUSSION TITLE
What are my options if I am running out GPU memory with Cone beam 3D reconstructions...

### DISCUSSION THREAD

**Discussion by nikhild1836** [astra-toolbox/astra-toolbox#366] [Q&A] [answered]
Hi all,
          I am trying to do 3D cone beam reconstructions with 2880 projections with 3040x3040 pixels in each projection. I am using an NVIDIA RTX A6000 GPU with 48 GB of RAM. 

         I have been able to do FDK_CUDA reconstructions at full resolution in less than a minute so kudos to the ASTRA team for this great toolbox which can handle such a large dataset!

          What I would like to do next is to do the same reconstructions using the CGLS3D_CUDA algorithm  So far I have had to down sample my projections to 760x760 but I would like to do this at full resolution. I run out of memory trying to do this.

          I came across this thread which suggests that there might be some options using the host memory: https://github.com/astra-toolbox/astra-toolbox/issues/161 (This is four years old, and no answer.)

          Elsewhere there were suggestions that plugins might be the way to solve this but I have not found any concrete solutions.

          Is there a solution to this problem either using host memory or shared memory or **multiple GPUs**? I have been able to use multiple GPUs for the FDK_CUDA algorithm, but CGSL3D does not seem to be able to use multiple GPUs.

Thank you!
Nikhil.

**Comment by wjp** [comment, marked as answer]
We have an alternative implementation of CGLS that uses host memory for the CGLS algorithm and the GPU(s) 
for the FP and BP operations. This also lets you use multiple GPUs. You can use it by calling

```
astra.plugin.register(astra.plugins.CGLSPlugin)
```

and then using the `CGLS-PLUGIN` algorithm (instead of `CGLS3D_CUDA`).

**Reply by nikhild1836** [reply]
Thank you for the quick reply. I tried that and got an error. See below. I am using ASTRA version: '2.1.2'
Any help is much appreciated!
---------------
`
Algorithm: CGLS-PLUGIN
iterations: 25
Error: Traceback (most recent call last):
  File "/home/userxxx/anaconda3/envs/astra/lib/python3.9/site-packages/astra/plugin.py", line 69, in astra_init
    self.initialize(cfg, *args, **kwargs)
  File "/home/userxxx/anaconda3/envs/astra/lib/python3.9/site-packages/astra/plugins/cgls.py", line 37, in initialize
    self.W = astra.OpTomo(cfg['ProjectorId'])
KeyError: 'ProjectorId'
.
.
  File "/home/userxxx/anaconda3/envs/astra/lib/python3.9/site-packages/astra/algorithm.py", line 36, in create
    return a.create(config)

  File "astra/algorithm_c.pyx", line 68, in astra.algorithm_c.create

Exception: Unable to initialize Algorithm.
`
--------------------

**Reply by wjp** [reply]
It's generally hard to debug from an error message without the corresponding script, but in this case it sounds like you're not passing the field 'ProjectorId' when creating the algorithm.

Try

```
projector_id = astra.create_projector('cuda3d',proj_geom,vol_geom)
cfg['ProjectorId'] = projector_id
```

**Reply by nikhild1836** [reply]
That seems to have started the reconstruction process. The missing line with the ProjectorId was not needed for the the CGLS3D algorithm.

Is there some example code for the CGLS-PLUGIN? In case there are some options related it that I might need later?

Thank you very much for the quick help!
Nikhil.
